pymeasure/instruments/thyracont/smartline_v2.py

```diff
@@ -0,0 +1,485 @@
+              '_SEDIS': "Sensor element disabled."}
+
+    def __init__(self, adapter, name="Thyracont SmartlineV2 Transmitter", baud_rate=115200,
+                 address=1, timeout=250,
+                 **kwargs):
+        super().__init__(adapter, name=name, includeSCPI=False,
+                         write_termination="\r",
+                         read_termination="\r",
+                         timeout=timeout,
+                         asrl={'baud_rate': baud_rate,
+                               'parity': Parity.none,
+                               'stop_bits': StopBits.one,
+                               },
+                         **kwargs
+                         )
+        self.address = address  # 1-16
+
+    def write(self, command):
+        """Write a command to the device."""
+        message = f"{self.address:03}{command}"
+        super().write(f"{message}{calculate_checksum(message)}")
+
+    def write_composition(self, accessCode, command, data=""):
+        """Write a command with an accessCode and optional data to the device.
+
+        :param accessCode: How to access the device.
+        :param command: Two char command string to send to the device.
+        :param data: Data for the command.
+        """
+        self.write(f"{accessCode}{command}{compose_data(data)}")
+
+    def ask(self, command_message, query_delay=0):
+        """Ask for some value and check that the response matches the original command.
+
+        :param str command_message: Access code, command, length, and content.
+            The command sent is compared to the response command.
+        """
+        self.write(command_message)
+        self.wait_for(query_delay)
+        return self.read(command_message[1:3])
+
+    def ask_manually(self, accessCode, command, data="", query_delay=0):
+        """
+        Send a message to the transmitter and return its answer.
+
+        :param accessCode: How to access the device.
+        :param command: Command to send to the device.
+        :param data: Data for the command.
+        :param int query_delay: Time to wait between writing and reading.
+        :return str: Response from the device after error checking.
+        """
+        self.write(f"{accessCode}{command}{compose_data(data)}")
+        self.wait_for(query_delay)
+        return self.read(command)
+
+    def read(self, command=None):
+        """Read from the device and do error checking.
+
+        :param str command: Original command sent to the device to compare it with the response.
+            None deactivates the check.
+        """
+        # Sometimes the answer contains 0x00 or values above 127, such that
+        # decoding fails.
+        response = self.read_bytes(-1, break_on_termchar=True)
+        if b"\x00" in response:
+            response = response.replace(b"\x00", b"")
+
+        # b"\r" is the termination character
+        got = response.rstrip(b"\r").decode('ascii', errors="ignore")
```

Comment:
```suggestion
        got = super().read().replace('\x00', '')
```
I propose this suggestion, it would do the conversion to string for you, but I am not sure what `pyvisa` does when a char > 127 is received. I have seen tests failing, for two reasons:

1. Errors in decoding are not ignored (https://github.com/pymeasure/pymeasure/blob/55e453cc33fe4aa2410389cc8bef779fc8d2222f/pymeasure/adapters/protocol.py#L135). Should be ignored ?
2. `ProtocolAdapter` does not implement `read_termination`. Is this correct ?

Reply:
> but I am not sure what pyvisa does when a char > 127 is received

that is exactly the issue, why I implemented that complicated method: pyvisa uses decode, which hangs with byte values above 127.

> ProtocolAdapter does not implement read_termination. Is this correct ?

The ProtocolAdapter will not break at the TermChar, that is correct.

Reply:
Reasons for the elements:
1. `rstrip`: If we use read_bytes until a term_char, the term_char will be returned as well, so we have to strip it (any adapter, be it serial, VISA or protocol)
2. ignore errors: to skip those extra chars, which happen to end up in the communication.


Btw, it could be a problem with my RS485 adapter, but I don't have an alternative to test.

Reply:
Ok, I agree this is weird, but I am sorry I have no experience on RS485 either, so I am not able to comment.
At least, the `if` in line 324 can be removed and do always the replace.

Off topic: I am not familiar about the `ProtocolAdapter`, but the behaviour is not consistent with `SerialAdapter`, but I understand this is intentional. Probably I need to read the documentation before commenting further.

Reply:
> Off topic: I am not familiar about the ProtocolAdapter, but the behaviour is not consistent with SerialAdapter, but I understand this is intentional. Probably I need to read the documentation before commenting further.

The behaviour is consistent, except for the "break_on_termchar". The reason is, that for the `ProtocolAdapter` you do not specify term_chars (to be agnostic to term_chars), therefore, the Adapter cannot recognize a termchar and cannot stop there.
However, you give the communication to the Adapter, so reading two times up to a termchar would require the following `comm__pairs = [(None, b"part one"), (None, b"part two")]` for an instrument returning `b"part one\npart two\n"` with newline as term_char.